trainer: migration guide from training operator to Trainer V2 (#4142)
kubeflow/website: content/en/docs/components/trainer/operator-guides/migration.md

- Enhanced MPI support: featuring MPI-Operator V2 features with SSH-based optimization to boost
  MPI performance.

> ## Migration paths

Reviewer: We can keep section titles capitalised. 

```suggestion
## Migration Paths
```

Kubeflow Trainer v2 introduces new APIs that replace the older, framework-specific CRDs such as `PyTorchJob`, `TFJob`, and `MPIJob`. These new APIs—`TrainingRuntime`, `TrainJob`, and `ClusterTrainingRuntime`—offer a more flexible and unified interface for defining training jobs across frameworks.

> This section shows how to migrate an existing PyTorch job to the new API.

Reviewer: ```suggestion
This section shows how to migrate an existing PyTorchJob to the new API.
```

> ### Migrate PyTorchJob to TrainingRuntime

Reviewer: I think, this example should be converted differently.
We should use [the default Torch runtime](https://github.com/kubeflow/trainer/blob/master/manifests/base/runtimes/torch_distributed.yaml), and use TrainJob to override:
```
image:
command:
numNodes
```

You should explain that Kubeflow Trainer also provides Python SDK to allow ML Users submit TrainJob without `kubectl`.

Author: I’ve updated the migration guide with a TrainJob configuration, using the default Torch runtime, where MPI isn’t needed for this torchrun-based example since it relies on PyTorch’s native backends (e.g., NCCL/Gloo). I’ve included replicas: 2 to override the number of nodes, as you suggested, and removed the mpiPolicy to simplify the setup. The updated configuration is as follows:

<pre> 
apiVersion: trainer.kubeflow.org/v2alpha1
kind: TrainJob
metadata:
  name: pytorch-simple
  namespace: kubeflow
spec:
  runtimeRef:
    name: torch-distributed 
  template:
    spec:
      containers:
        - name: trainer
          image: docker.io/kubeflowkatib/pytorch-mnist:v1beta1-45c5727
          command: ["torchrun", "/opt/pytorch-mnist/mnist.py", "--epochs=1"]
          env:
            - name: MASTER_ADDR
              value: "pytorch-simple-0"
            - name: MASTER_PORT
              value: "29400"
      replicas: 2
</pre>

I’ve also added a note explaining that Kubeflow Trainer provides a Python SDK, allowing ML users to submit TrainJob workloads programmatically without Kubectl, as recommended. Please let me know if you have further suggestion to improve.

Reviewer: You should use [.spec.trainer.numNodes API](https://github.com/kubeflow/trainer/blob/master/pkg/apis/trainer/v1alpha1/trainjob_types.go#L222) and remove the MASTER_ADDR, MASTER_PORT from env vars, since these values are set by controller automatically.

The following example demonstrates how to migrate from a `PyTorchJob` (v1) to a `TrainingRuntime` (v2alpha1).

### Old: PyTorchJob (v1)
```yaml
apiVersion: kubeflow.org/v1
kind: PyTorchJob
metadata:
  name: pytorch-simple
  namespace: kubeflow
spec:
  pytorchReplicaSpecs:
    Master:
      replicas: 1
      restartPolicy: OnFailure
      template:
        spec:
          containers:
            - name: pytorch
              image: docker.io/kubeflowkatib/pytorch-mnist:v1beta1-45c5727
              command: ["python3", "/opt/pytorch-mnist/mnist.py", "--epochs=1"]
    Worker:
      replicas: 1
      restartPolicy: OnFailure
      template:
        spec:
          containers:
            - name: pytorch
              image: docker.io/kubeflowkatib/pytorch-mnist:v1beta1-45c5727
              command: ["python3", "/opt/pytorch-mnist/mnist.py", "--epochs=1"]
```

> ### New: TrainingRuntime (v2alpha1)
> ```yaml
> apiVersion: trainer.kubeflow.org/v2alpha1
> kind: TrainingRuntime

Reviewer: Let's use ClusterTrainingRuntime as a default migration path, since it is the default runtimes to use.
```suggestion
### New: ClusterTrainingRuntime (v2alpha1)

```yaml
apiVersion: trainer.kubeflow.org/v2alpha1
kind: ClusterTrainingRuntime
```

metadata:
  name: torch-distributed-multi-node
spec:
  mpiPolicy:
    numNodes: 2
  template:
    spec:
      replicatedJobs:
        - name: node
          template:
            metadata:
              labels:
                trainer.kubeflow.org/trainjob-ancestor-step: trainer
            spec:
              containers:
                - name: trainer
                  image: docker.io/kubeflowkatib/pytorch-mnist:v1beta1-45c5727
                  env:
                    - name: MASTER_ADDR
                      value: "pytorch-node-0-0.pytorch"
                    - name: MASTER_PORT
                      value: "29400"
                  command: ["torchrun", "train.py"]
```

> The new `TrainingRuntime` CRD allows you to define reusable templates and use MPI-based distributed training with greater flexibility. The `replicatedJobs` and `mpiPolicy` fields replace framework-specific roles like `Master` and `Worker`.

Reviewer: MPI is used only for ML frameworks that require MPI functionality (e.g. DeepSpeed). So it is not needed for `torchrun`.

### Version compatibility table

The following table lists compatibility between Kubeflow Trainer versions, supported APIs, and minimum Kubernetes versions:

| Trainer version | API version | CRDs introduced                                | Minimum Kubernetes version | Notes                        |
|-----------------|-------------|-------------------------------------------------|-----------------------------|-------------------------------|
| v1.4–v1.8       | v1          | `PyTorchJob`, `TFJob`, `MPIJob`, etc.          | 1.23+                       | Training Operator style       |

> | v1.9+           | v2alpha1    | `TrainingRuntime`, `TrainJob`, `ClusterTrainingRuntime` | 1.31.3+                     | Trainer v2 (breaking changes) |

Reviewer: This table is incorrect, since we don't support TrainJob in Training Operator 1.9.
I would suggest remove it.

### Additional information

- Trainer v2 does not use individual CRDs per framework. Instead, it abstracts functionality into a unified runtime specification.

> - You can use the `Kubeflow Python SDK` to generate and submit Trainer v2 workloads programmatically.

Reviewer: We should explain that Kubeflow Python SDK is the primary interface for ML Users to interact with TrainJob APIs.

- Initializer and multi-node coordination logic is managed by the system, reducing overhead on the training container.

> - Support for specifying `managedBy` fields and runtime API versions (via `runtimeRef.version`) is under consideration.

Reviewer: What do you mean by that ?

> You can find detailed proposals and discussion [in the Kubeflow Trainer GitHub repository](https://github.com/kubeflow/trainer/tree/master/docs/proposals).

Reviewer: Please add link to the KEP: https://github.com/kubeflow/trainer/tree/master/docs/proposals/2170-kubeflow-trainer-v2
